operations/ls.go

```diff
@@ -2,37 +2,166 @@ package operations
 
 import (
 	"errors"
+	"runtime"
+	"strings"
+
+	"github.com/gobwas/glob"
 
 	"go.beyondstorage.io/v5/pairs"
 	"go.beyondstorage.io/v5/types"
 )
 
 func (so *SingleOperator) List(path string) (ch chan *ObjectResult, err error) {
-	it, err := so.store.List(path, pairs.WithListMode(types.ListModeDir))
-	if err != nil {
-		return nil, err
-	}
-
 	ch = make(chan *ObjectResult, 16)
 	go func() {
 		defer close(ch)
 
-		for {
-			o, err := it.Next()
-			if err != nil && errors.Is(err, types.IterateDone) {
-				break
-			}
-			if err != nil {
-				ch <- &ObjectResult{Error: err}
-				break
-			}
+		objects, err := so.GlobList(path)
+		if err != nil {
+			ch <- &ObjectResult{Error: err}
+			return
+		}
+
+		for _, object := range objects {
+			if object.Mode.IsDir() {
+				it, err := so.store.List(object.Path, pairs.WithListMode(types.ListModeDir))
+				if err != nil {
+					ch <- &ObjectResult{Error: err}
+					break
+				}
 
-			ch <- &ObjectResult{Object: o}
+				for {
+					o, err := it.Next()
+					if err != nil && errors.Is(err, types.IterateDone) {
+						break
+					}
+					if err != nil {
+						ch <- &ObjectResult{Error: err}
+						break
+					}
+
+					ch <- &ObjectResult{Object: o}
+				}
+			} else {
+				ch <- &ObjectResult{Object: object}
+			}
 		}
 	}()
 
 	return ch, nil
 }
+
+func hasMeta(path string) bool {
+	magicChars := `*?[{`
+	if runtime.GOOS != "windows" {
+		magicChars = `*?[{\`
+	}
+	return strings.ContainsAny(path, magicChars)
+}
+
+func splitPath(path string) (prefixPath, pattern string) {
+	if path == "" {
+		return
+	}
+
+	subpaths := strings.Split(path, "/")
+	for _, subpath := range subpaths {
+		if hasMeta(subpath) {
+			pattern = path[len(prefixPath):]
+			break
+		} else {
+			prefixPath += subpath + "/"
+		}
+	}
+
+	prefixPath = strings.TrimSuffix(prefixPath, "/")
+
+	return
+}
+
+func (so *SingleOperator) globList(path, pattern string) (objects []*types.Object, err error) {
+	var g glob.Glob
+	if pattern != "" {
+		g = glob.MustCompile(pattern)
+	}
+
+	it, err := so.store.List(path, pairs.WithListMode(types.ListModeDir))
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		o, err := it.Next()
+		if err != nil && errors.Is(err, types.IterateDone) {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if (g != nil && g.Match(o.Path)) || g == nil {
+			objects = append(objects, o)
+		}
+	}
+
+	return
+}
+
+func (so *SingleOperator) DirWalk(workPath string, pattern string) (objects []*types.Object, err error) {
+	if pattern == "" {
+		objs, err := so.globList(workPath, "")
+		if err != nil {
+			return nil, err
+		}
+		if len(objs) == 1 && !objs[0].Mode.IsDir() {
+			objects = append(objects, objs...)
+		} else {
+			objects = append(objects, &types.Object{
+				Mode: types.ModeDir,
+				Path: workPath,
+				ID:   workPath,
+			})
+		}
+
+		return objects, nil
+	}
```

Comment:
I'm confused about whether the `GlobList` (to expand glob pattern) should return `objects []*types.Object` or `files []string`. If returns `[]string`, we need to repeat the `list`. If returns `[]*types.Object`, the above processing looks strange.

Reply:
It does look strange.

How about keeping `List` unchanged, and implementing a `ListWithGlob(path, pattern string)`?

In `ListWith`

Reply:
Why we need a `DirWalk` here?

Reply:
I'm trying to recursively expand the pattern in `DirWalk`.